src/streamers/__tests__/spotify.streamer.spec.ts

```diff
@@ -0,0 +1,483 @@
+      const payload = { access_token: fakeAccessToken }
+      storer.set(SPOTIFY_ACCESS_TOKEN_KEY, payload, 5000)
+      const response = await streamer.getAccessToken(refreshToken)
+      expect(response.access_token).toBe(fakeAccessToken)
+    })
+
+    test('fetch from spotify if not available in cache (and save in cache afterwards', async () => {
+      const response = await streamer.getAccessToken(refreshToken)
+      expect(response.access_token).toBe(freshAccessToken)
+
+      const storedToken = storer.get<SpotifyAccessToken>(
+        SPOTIFY_ACCESS_TOKEN_KEY
+      )
+      expect(storedToken).toEqual(response)
+    })
+
+    test('throw error if token fetching is unsuccessful', async () => {
+      expect.assertions(1)
+      return streamer.getAccessToken('fake-refresh-token').catch((error) => {
+        expect(error).toEqual({
+          message: 'something went wrong',
+          status: 400,
+        })
+      })
+    })
+  })
+
+  describe('fetchLastPlayed()', () => {
+    const streamer = new SpotifyStreamer(storer, {
+      clientId,
+      clientSecret,
+      refreshToken,
+      useCache,
+      cacheDuration,
+    })
+
+    beforeEach(() => {
+      streamer.setUseCache(true)
+      storer.clear()
+    })
+
+    beforeAll(() => {
+      const fakeAccessToken = "bolaji's fake access token"
+      const payload = { access_token: fakeAccessToken }
+      storer.set(SPOTIFY_ACCESS_TOKEN_KEY, payload, 5000)
+    })
+    afterAll(() => {
+      streamer.setUseCache(true)
+    })
+
+    test('should return an error if http request errors', async () => {
+      expect.assertions(1)
+      return streamer
+        .fetchLastPlayed({
+          access_token: accessTokenForError,
+        } as SpotifyAccessToken)
+        .catch((error) => {
+          expect(error).toEqual({
+            message: 'something went wrong',
+            status: 400,
+          })
+        })
+    })
+
+    test('should return null if endpoint doesnt contain item', async () => {
+      const response = await streamer.fetchLastPlayed({
+        access_token: accessTokenForNull,
+      } as SpotifyAccessToken)
+      expect(response).toBeNull()
+    })
+
+    test('should return the last played song and not cache if useCache is false', async () => {
+      streamer.setUseCache(false)
+      const response = await streamer.fetchLastPlayed({
+        access_token: freshAccessToken,
+      } as SpotifyAccessToken)
+      const cachedSong = storer.get(SPOTIFY_TRACK_KEY)
+      expect(cachedSong).toBeUndefined()
+      expect(response).toEqual({
+        is_playing: false,
+        title: 'Coming Home',
+        artiste: 'Kanye West',
+        image_url: 'https://i.scdn.co/image/ab67616d0000b273712549143',
+        preview_url:
+          'https://p.scdn.co/mp3-preview/0663627c27885467868491a91185643b0508975c?cid=774b29d4f13844c495f206cafdad9c86',
+        url: 'https://open.spotify.com/track/6M2wZ9GZgrQXHCFfjv46we',
+      })
```

Comment:
I think we can move this and others to a data file, it is repeated in quite a number of places.

Reply:
Yeah, I'll create a fixture file for this.

Reply:
Done